### PR TITLE
perf: optimize GLM-5.2 serving

### PR DESCRIPTION
**Pull request by CUHKSZzxy**
## Summary

- Optimize GLM-5.2 FP8 DSA scoring with contiguous DeepGEMM MQA logits for prefill and paged DeepGEMM MQA logits for decode.
- Fuse DSA indexer preprocessing, noaux routing, and Q/KV-A projections; use compact blocked-FP8 MoE at low route density.
- Shard the LM-head vocabulary across tensor-parallel ranks and fix TP-local loading of merged FP8 scales.
- Run DeepSeek-V3.2 MTP with sparse DSA and synchronize the finalized speculative-model config with the engine.

## Performance

GLM-5.2 FP8 TP8 on 8×H200 GPUs with DeepGEMM, FP8 KV cache, MTP5, concurrency 16, and five-turn SWE-Smith:

| Version | Completion throughput | Steady throughput | TPOT | Success |
| --- | ---: | ---: | ---: | ---: |
| GLM-5.2 support branch | 678.74 tok/s | 1,137.82 tok/s | 18.14 ms | 80/80 |
| Current optimization stack | **847.84 tok/s** | **1,289.04 tok/s** | **14.83 ms** | 80/80 |
| Change | **+24.91%** | **+13.29%** | **-18.25%** | unchanged |

Both results are profiler-free runs using the same server configuration and live SWE-Smith workload. This is a cumulative full-branch comparison rather than attribution to any individual optimization.

## Validation

- Pre-commit checks passed.
- 149 targeted attention, cache, DeepGEMM, FP8 MoE, routing, projection, embedding, and NSA/MTP tests passed.
- TP8 model loading, CUDA-graph warmup, repeated SWE runs, and eight-rank timeline capture completed successfully.
- Current-stack full GSM8K: **1,266/1,319 (95.98%)**, with no request, empty-response, or answer-extraction failures.

## Assistance

Assisted with Codex + GPT-5.6-Sol xHigh, reviewed manually

### CODE REVIEW

File: lmdeploy/pytorch/nn/embedding.py [hunk 1]
```diff
@@ -107,3 +108,68 @@ def forward(self, x: torch.Tensor):
...
+        input_size = local_logits.size()
+        output_size = (input_size[0] * self.tp, ) + input_size[1:]
+        logits = local_logits.new_empty(output_size)
+        dist.all_gather_into_tensor(logits, local_logits, group=self.tp_group)
+        # The collective concatenates dim 0. Move its rank dimension beside
+        # the vocabulary shard before reconstructing the full last dimension.
+        logits = logits.reshape((self.tp, ) + input_size)
+        logits = logits.movedim(0, local_logits.dim() - 1)
+        logits = logits.reshape(input_size[:-1] + (self.tp * input_size[-1], ))
+        return logits[..., :self.vocab_size]
```
[the remark this answers is not in the document]
Comment (CUHKSZzxy) [reply lmdeploy/pytorch/nn/embedding.py:170]: The dim-0 concatenation is intentional. After the collective, we reshape `[tp * ..., vocab_shard]` to `[tp, ..., vocab_shard]`, move the TP axis beside the last dimension, then flatten `tp * vocab_shard` into the full vocabulary. This preserves the batch/token dimensions for both 2-D and 3-D logits. I also verified the real NCCL path with `tests/pytorch/nn/test_embedding.py::test_parallel_lm_head` on two GPUs.

File: tests/pytorch/nn/test_embedding.py [hunk 2]
```diff
@@ -123,3 +150,35 @@ def test_embedding(self, vocab_size, feat_size, padding_idx, seqlen, tp, dtype,
                     p.kill()
 
         torch.testing.assert_close(out, gt)
+
+
+def test_parallel_lm_head():
```
[the remark this answers is not in the document]
Comment (CUHKSZzxy) [reply tests/pytorch/nn/test_embedding.py:160]: Fixed in fc238397 with a module-level skip requiring at least two visible CUDA devices, since the existing TP embedding test has the same TP=2 NCCL requirement. Verified with one visible GPU (19 skipped) and two visible GPUs (19 passed).

File: lmdeploy/pytorch/backends/nsa.py [hunk 3]
```diff
@@ -11,23 +12,82 @@ class NSAIndexMeta:
...
+            cu_seqlens_q[:-1], q_seqlens, output_size=num_tokens)
+        history_lengths = torch.repeat_interleave(
+            kv_seqlens - q_seqlens, q_seqlens, output_size=num_tokens)
+        query_offsets = torch.arange(
+            num_tokens, device=q_seqlens.device, dtype=q_start.dtype) - q_start
+        indexer_kv_seqlens = history_lengths + query_offsets + 1
+    return indexer_kv_seqlens.to(torch.int32)
+
+
+def build_nsa_index_meta(*, num_tokens: int, is_decoding: bool,
```
Comment (grimoire) [inline lmdeploy/pytorch/backends/nsa.py:43]: Build nsa index meta would be called every dsa layer, right? so repeat_interleave and arange would be called multiple times.

Comment (CUHKSZzxy) [reply lmdeploy/pytorch/backends/nsa.py:43]: `build_nsa_index_meta()` is part of the per-step/CUDA-graph metadata preparation. All DSA layers share the same deduplicated metadata group,  so it runs once per metadata-preparation phase, and all layers reuse the result

**Review by grimoire — APPROVED** [review]
LGTM
